MercurialExtensions/fixutf8/win32helper.py

```diff
@@ -113,6 +242,212 @@ def getargs():
+					inherited by the child process.
+					"""
+					handles = HANDLE(), HANDLE()
+					if not CreatePipe(ctypes.byref(handles[0]),
+									  ctypes.byref(handles[1]), ctypes.byref(saAttr), 0):
+						raise ctypes.WinError()
+					if not SetHandleInformation(handles[uninherit],
+												HANDLE_FLAG_INHERIT, 0):
+						raise ctypes.WinError()
+					return handles[0].value, handles[1].value
+
+				p_hstdout, c_hstdout = create_pipe(uninherit=0)
+				# 'mergeout' signals that stdout and stderr should be merged.
+				# We do that by using one pipe for both of them.
+				if mergeout:
+					c_hstderr = HANDLE()
+					if not DuplicateHandle(GetCurrentProcess(), c_hstdout,
+										   GetCurrentProcess(), ctypes.byref(c_hstderr),
+										   0, True, DUPLICATE_SAME_ACCESS):
+						raise ctypes.WinError()
+				else:
+					p_hstderr, c_hstderr = create_pipe(uninherit=0)
+
+				# Create the process object
+				piProcInfo = PROCESS_INFORMATION()
+				siStartInfo = STARTUPINFO()
+				siStartInfo.cb = ctypes.sizeof(siStartInfo)
+				siStartInfo.hStdOutput = c_hstdout
+				siStartInfo.hStdError = c_hstderr
+				siStartInfo.dwFlags = STARTF_USESTDHANDLES
+				dwCreationFlags = CREATE_SUSPENDED | CREATE_NO_WINDOW  # | CREATE_NEW_CONSOLE
+
+				if not CreateProcess(None,
+									 cmd,
+									 None, None, True, dwCreationFlags,
+									 None, None, ctypes.byref(siStartInfo),
+									 ctypes.byref(piProcInfo)):
+					raise ctypes.WinError()
+
+				# Close this process's versions of the child handles
+				CloseHandle(c_hstdout)
+				c_hstdout = None
+				if c_hstderr is not None:
+					CloseHandle(c_hstderr)
+					c_hstderr = None
+
+				# Transfer ownership of the parent handles to the object
+				self.hstdout = p_hstdout
+				p_hstdout = None
+				if not mergeout:
+					self.hstderr = p_hstderr
+					p_hstderr = None
+				self.piProcInfo = piProcInfo
+
+			finally:
+				if p_hstdout:
+					CloseHandle(p_hstdout)
+				if c_hstdout:
+					CloseHandle(c_hstdout)
+				if p_hstderr:
+					CloseHandle(p_hstderr)
+				if c_hstderr:
+					CloseHandle(c_hstderr)
+
+			return self
+
+
+		def _stdout_thread(self, handle, func):
+			# Allocate the output buffer
+			data = ctypes.create_string_buffer(4096)
+			while True:
+				bytesRead = DWORD(0)
+				if not ReadFile(handle, data, 4096,
+								ctypes.byref(bytesRead), None):
+					le = GetLastError()
+					if le == ERROR_BROKEN_PIPE:
+						return
+					else:
+						raise ctypes.WinError()
+				# FIXME: Python3
+				s = data.value[0:bytesRead.value]
+				# print("\nv: %s" % repr(s), file=sys.stderr)
+				func(fromunicode(s))
+
+		def run(self, stdout_func=None, stderr_func=None):
+			"""Runs the process, using the provided functions for I/O.
+			The functions stdout_func and stderr_func are called whenever
+			something is printed to stdout or stderr, respectively.
+			These functions are called from different threads (but not
+			concurrently, because of the GIL).
```

Comment:
Yes, the longer name would not help. What I think I would most want to know is whether the reason is something clever we do somewhere (that a change to our code might break) or something inherent in Python. Something like "because python keeps this sort of thing sequential using a global lock (GIL)". You may know enough to be more specific than "this sort of thing" without being too verbose.